src/TRestGeant4Metadata.cxx

```diff
@@ -1443,40 +1484,40 @@ void TRestGeant4Metadata::ReadParticleSource(TiXmlElement* definition) {
     // Angular distribution parameters
     TiXmlElement* angularDefinition = GetElement("angularDist", sourceDefinition);
 
-    source.SetAngularDistType(GetFieldValue("type", angularDefinition));
+    source->SetAngularDistType(GetFieldValue("type", angularDefinition));
 
-    if (source.GetAngularDistType() == "TH1D") {
-        source.SetAngularFilename(SearchFile(GetFieldValue("file", angularDefinition)));
-        source.SetAngularName(GetFieldValue("spctName", angularDefinition));
+    if (source->GetAngularDistType() == "TH1D") {
+        source->SetAngularFilename(SearchFile(GetFieldValue("file", angularDefinition)));
+        source->SetAngularName(GetFieldValue("spctName", angularDefinition));
     }
 
-    if (fPrimaryGenerator.GetNumberOfSources() == 0 && source.GetAngularDistType() == "backtoback") {
+    if (GetNumberOfSources() == 0 && source->GetAngularDistType() == "backtoback") {
         cout << "WARNING: First source cannot be backtoback. Setting it to isotropic" << endl;
-        source.SetAngularDistType("isotropic");
+        source->SetAngularDistType("isotropic");
     }
 
-    source.SetDirection(StringTo3DVector(GetFieldValue("direction", angularDefinition)));
+    source->SetDirection(StringTo3DVector(GetFieldValue("direction", angularDefinition)));
 
     // Energy distribution parameters
     TiXmlElement* energyDefinition = GetElement("energyDist", sourceDefinition);
 
-    source.SetEnergyDistType(GetFieldValue("type", energyDefinition));
+    source->SetEnergyDistType(GetFieldValue("type", energyDefinition));
 
-    if (source.GetEnergyDistType() == "TH1D") {
-        source.SetSpectrumFilename(SearchFile(GetFieldValue("file", energyDefinition)));
-        source.SetSpectrumName(GetFieldValue("spctName", energyDefinition));
+    if (source->GetEnergyDistType() == "TH1D") {
+        source->SetSpectrumFilename(SearchFile(GetFieldValue("file", energyDefinition)));
+        source->SetSpectrumName(GetFieldValue("spctName", energyDefinition));
     }
 
-    source.SetEnergyRange(Get2DVectorParameterWithUnits("range", energyDefinition));
+    source->SetEnergyRange(Get2DVectorParameterWithUnits("range", energyDefinition));
 
-    if (source.GetEnergyDistType() == "mono") {
+    if (source->GetEnergyDistType() == "mono") {
         Double_t en;
         en = GetDblParameterWithUnits("energy", energyDefinition);
-        source.SetEnergyRange(TVector2(en, en));
-        source.SetEnergy(en);
+        source->SetEnergyRange(TVector2(en, en));
+        source->SetEnergy(en);
     }
 
-    fPrimaryGenerator.AddSource(source);
+    AddParticleSource(source);
```

Comment:
Perhaps `source` cannot be deleted because of this line?

Would be better to move to `unique_ptr`?